Feature/registration form

mangose10 opened the pull request:
Whole functionality of register form. GUI errors need to be made more presentable, but the base is there.

Commit:
Test push

mangose10 committed:
registrationForm

Create a front end that adds these forms for the registration:
-----------------------------------------------------------------
i. Email address
ii. Password
iii. Display Name
iv. Institution
v. Website
*You need to follow the same frontend as the login page*

Validate and show errors on the following forms on the client side:
---------------------------------------------------
i. Email address (should be validated as an email address and should be made
lower-case before insertion in to the database.
ii. Password (should be validated with a password confirmation, should be no less
than 8 characters long, and should contain both upper case, lower case, and at
least one numeric digit).
iii. Display Name (should be something other than their email)

Validate the registration forms once it gets to the server side:
---------------------------------
-Email address for uniqueness(It cannot repeat on the database)
-Display name(must be unique)
Show the error back in the client side:
--------------------------
-Say that the email is already registered
-Say that the display name is already registered

Dependencies:
---------------
-Both validation must be done(on client side and server side)

Insert all of the new data from registration form into the MongoDB once all of it has been validate it:
----------------------------------------------------------------------------
-Email address (should be made lower-case before insertion in to the database.)
-Password
-Display Name
-Institution (optional)
-Website (optional)

mangose10 committed:
registrationForm

Create a front end that adds these forms for the registration:
-----------------------------------------------------------------
i. Email address
ii. Password
iii. Display Name
iv. Institution
v. Website
*You need to follow the same frontend as the login page*

Validate and show errors on the following forms on the client side:
---------------------------------------------------
i. Email address (should be validated as an email address and should be made
lower-case before insertion in to the database.
ii. Password (should be validated with a password confirmation, should be no less
than 8 characters long, and should contain both upper case, lower case, and at
least one numeric digit).
iii. Display Name (should be something other than their email)

Validate the registration forms once it gets to the server side:
---------------------------------
-Email address for uniqueness(It cannot repeat on the database)
-Display name(must be unique)
Show the error back in the client side:
--------------------------
-Say that the email is already registered
-Say that the display name is already registered

Dependencies:
---------------
-Both validation must be done(on client side and server side)

Insert all of the new data from registration form into the MongoDB once all of it has been validate it:
----------------------------------------------------------------------------
-Email address (should be made lower-case before insertion in to the database.)
-Password
-Display Name
-Institution (optional)
-Website (optional)

Update auth.controller.ts

first set of review changes

first set only includes standard styling changes

first set of reviews pt.2

small changes that were missed in the first commit

second portion of RegisterForm review

This includes the password and email confirms.

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -18,6 +19,10 @@ export class AuthService {
     private readonly $configService: ConfigService
   ) {}
 
+  public E_ERR = 0;
```
Should be defined as a string Enum for error handling.

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -18,6 +19,10 @@ export class AuthService {
   ) {}
 
+  public E_ERR = 0;
+  public D_ERR = 1;
```
Should be defined as a string Enum for error handling

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -18,6 +19,10 @@ export class AuthService {
 
+  public E_ERR = 0;
+  public D_ERR = 1;
+  public B_ERR = 2;
```
Should be defined as a string Enum for error handling

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -44,8 +49,30 @@ export class AuthService {
+    if (website != ''){uci.website = website;}
+    if (institution != ''){uci.institution = institution;}
+
+    return this.registerGraphQL(uci, password);
```
This doesn't need to be separated yet.

mangose10 replied:
I changed this to:

if (true /*website != ''*/){uci.website = website;}
if (true /*institution != ''*/){uci.institution = institution;}

Is that fine?

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -86,14 +113,27 @@ export class AuthService {
         email: input.email
       }
     });
+    let checkD = await this.$databaseService.user.findUnique({
```
This can be combined into one query.

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -86,14 +113,27 @@ export class AuthService {
-    if (user) {
-      throw new UnauthorizedException('User already exists');
+    //NotUniqueException error throws
+    if (checkE && checkD){
```
This can be combined into one exception check

mangose10 replied:
This is what I changed it to:

try {
      let check = await this.$databaseService.user.findMany({
        where: {
          OR: [
            {email: input.email},
            {displayName: input.displayName}
          ]
        }
      });
      if (check.length > 1){
        throw UniqueError.B_ERR;
      }
      if (check[0].email === input.email){
        throw UniqueError.E_ERR;
      }
      if (check[0].displayName === input.displayName){
        throw UniqueError.D_ERR;
      }
    }catch(e){
      if (e != {}){
        throw new NotUniqueException(e);
      }
    }
    
    ------------------------------------------------
 I couldn't find a way to make a custom exception for PrismaClient, and it normally catches a "Not Found" exception (which is what we want in order to proceed) so this is what I went with

quinnjr replied:
It doesn't need a try-catch block and should really just use HTTP statuses to send the error message.

quinnjr commented on src/app/auth/register/register.component.scss:
```diff
@@ -0,0 +1,12 @@
+.reqRed {
```
This is defined in the bulma css

quinnjr commented on src/app/auth/register/register.component.scss:
```diff
@@ -0,0 +1,12 @@
+  color: green;
+}
+
+.invis {
```
This is defined in the bulma css

quinnjr commented on src/app/auth/register/register.component.scss:
```diff
@@ -0,0 +1,12 @@
+}
+
+.greenAcpt {
+  color: green;
```
This is defined in the bulma css

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+    private readonly $fb: FormBuilder,
+    private readonly $http: HttpClient
+  ) {
+    this.registerForm = this.$fb.group({
```
Move the registration of the form group outside of the constructor

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+    }
+  }
+
+  public ngOnDestroy() {}
```
Don't need an empty method

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+  public ngOnDestroy() {}
+
+  public onSubmit() {
+    console.log(this.registerForm);
```
Only console log on debug mode

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+  public oldPass: string = '';
+  public oldEmail: string = '';
+  public oldDName: string = '';
+  public uniqueE: boolean = true;
```
This seems extrenous

mangose10 replied:
Bad naming sense from my part. Should be "uniqueEmailFlag". It gets flagged false if backend response returns with E_ERR or B_ERR. This is so that user knows why submission did not pass. Do you want me to do this a different way?

quinnjr replied:
Use HTTP statuses and messages from them.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+  public oldEmail: string = '';
+  public oldDName: string = '';
+  public uniqueE: boolean = true;
+  public uniqueD: boolean = true;
```
This seems extrenous

mangose10 replied:
Bad naming sense from my part. Should be "uniqueDisplayNameFlag". It gets flagged false if backend response returns with D_ERR or B_ERR. This is so that user knows why submission did not pass. Do you want me to do this a different way?

quinnjr replied:
Use HTTP statuses and error messages from them.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+
+    this.$http.post('/api/auth/register',
+    {
+      email:email,
```
Ensure email is sent as lower case

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+    const password = this.registerForm.get('password')?.value;
+
+    this.$http.post('/api/auth/register',
+    {
```
Use object syntactic sugar

quinnjr commented on src/app/navigation/navigation.component.html:
```diff
@@ -27,6 +27,7 @@
           <a class="navbar-item" routerLink="/admin/plugins">Plugin Management</a>
       </div>
     </div>
+    <a class="navbar-item" routerLink="/register">Register</a>
```
These should be in the same page

mangose10 replied:
I'm not understanding this part... I know you mentioned this in the meeting last time, but I thought you meant to just add this to the navbar when you said to put registerForm next to the login. 

Did you mean to have the login page render registerForm if necessary and loginForm if not?

quinnjr replied:
Nevermind. We'll just do two pages

mangose10 replied:
I feel like you gave up on me :(

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+  ) {
+    this.registerForm = this.$fb.group({
+      email: ['', [Validators.required, Validators.email]],
+      dName: ['', [Validators.required, sameAsEmailValidator("")]],
```
Should be displayName

Doesn't need to validate email, display name can be different

mangose10 replied:
My naming sense is horrible, I'll work on it. This is a custom validator that verifies that the Display Name is not the same as the email handle, will be switched to "notEmailValidator("")"

mangose10 replied:
After reading this comment I realized that you want a Validator that checks if an email handle is unique and another that checks if displayName is unique at every character input. In that case the uniqueEmailFlag and uniqueDisplayNameFlag are indeed extraneous.

Would I still need to validate in registerGQL() then? No right? I'd just need to create a new post in the auth.controller and a new function in auth.service? If so I might need to redo this whole part.

quinnjr replied:
You need to validate everything on the backend and use HTTP statuses to confirm validity.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+    private readonly $http: HttpClient
+  ) {
+    this.registerForm = this.$fb.group({
+      email: ['', [Validators.required, Validators.email]],
```
Needs a corresponding emailConfirm

quinnjr commented on server/auth/auth.controller.ts:
```diff
@@ -15,12 +16,28 @@ export class AuthController {
   ): Promise<any> {
-    return this.$authService.register(email, password);
+    return this.$authService.register(email, displayName, website, institution, password)
+    .catch(err => {return err.response});
```
Don't catch and allow NestJS to handle the errir

quinnjr commented on server/auth/auth.controller.ts:
```diff
@@ -1,9 +1,10 @@
+import { NotUniqueException } from "./not-unique.exception";
 import { LocalAuthGuard } from './local-auth.guard';
 import { Public } from '../public.decorator';
+import { catchError } from "rxjs";
```
Unnecessary import?

quinnjr commented on server/auth/auth.controller.ts:
```diff
@@ -1,9 +1,10 @@
 import { AuthGuard } from '@nestjs/passport';
 import { AuthService } from './auth.service';
-
+import { NotUniqueException } from "./not-unique.exception";
```
Unnecessary import?

quinnjr commented on server/auth/auth.controller.ts:
```diff
@@ -1,9 +1,10 @@
-import { Controller, Request, Post, UseGuards, Body } from '@nestjs/common';
+import { Controller, Request, Post, Get, UseGuards, Body, UseInterceptors } from '@nestjs/common';
```
Unnecessary UseInterceptors?

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -44,8 +49,30 @@ export class AuthService {
+   * @param institution   User institution (optional)
+   * @param password      User password, not hashed
+   * @returns HttpResponse Promise, will include
+   *          Success -> response.body will have user input data
```
Response body should include only a valid JWT

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -44,8 +49,30 @@ export class AuthService {
+    if (website != ''){uci.website = website;}
+    if (institution != ''){uci.institution = institution;}
+
+    return this.registerGraphQL(uci, password);
```
Should return a JWT

quinnjr commented on server/auth/not-unique.exception.ts:
```diff
@@ -0,0 +1,21 @@
+ *            2 -> both not unique
+ */
+export class NotUniqueException extends HttpException {
+  constructor(nui: any) {
```
Should be an enum

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -10,20 +10,96 @@ <h2>Register</h2>
-                [ngClass]="{ 'is-danger': isSubmitted && registerForm.get('emails')?.errors }" required />
+                [ngClass]="{
+                  'is-danger': isSubmitted && registerForm.get('email')?.errors,
+                  'reqRed': !registerForm.get('email')?.valid
```
Use Bulma css

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -10,20 +10,96 @@ <h2>Register</h2>
+          <div class="field">
+            <label class="label">Display Name</label>
+            <div class="control">
+              <input class="input" type="text" placeholder="bob"
```
Don't have a placeholder

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -10,20 +10,96 @@ <h2>Register</h2>
             <div class="control">
               <input class="input" type="email" placeholder="jdoe@fiu.edu"
```
Don't have a placeholder

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -10,20 +10,96 @@ <h2>Register</h2>
+          <div class="field">
+            <label class="label">Institution</label>
+            <div class="control">
+              <input class="input" type="institution" placeholder="FIU"
```
Don't have a placeholder

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -10,20 +10,96 @@ <h2>Register</h2>
+          <div class="field">
+            <label class="label">Website</label>
+            <div class="control">
+              <input class="input" type="website" placeholder="www.fiu.com"
```
Don't have a placeholder

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
+import { Component, OnDestroy, OnInit, OnChanges, DoCheck } from '@angular/core';
+import { FormGroup, FormBuilder, Validators } from '@angular/forms';
+import { sameAsEmailValidator } from "./verify/validate/same-as-email.directive";
+import { samePassValidator } from "./verify/validate/pass-match.directive";
```
https://stackoverflow.com/questions/51605737/confirm-password-validation-in-angular-6

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
-import { FormControl, FormGroup } from '@angular/forms';
+import { Component, OnDestroy, OnInit, OnChanges, DoCheck } from '@angular/core';
+import { FormGroup, FormBuilder, Validators } from '@angular/forms';
+import { sameAsEmailValidator } from "./verify/validate/same-as-email.directive";
```
https://stackoverflow.com/questions/51605737/confirm-password-validation-in-angular-6

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,22 +1,104 @@
   public isSubmitted: boolean = false;
+  public oldPass: string = '';
+  public oldEmail: string = '';
+  public oldDName: string = '';
```
oldDisplayName

mangose10 commented on server/auth/auth.controller.ts:
```diff
@@ -1,9 +1,10 @@
-import { Controller, Request, Post, UseGuards, Body } from '@nestjs/common';
+import { Controller, Request, Post, Get, UseGuards, Body, UseInterceptors } from '@nestjs/common';
 import { AuthGuard } from '@nestjs/passport';
 import { AuthService } from './auth.service';
```
Test comment

mangose10 commented:
Test

quinnjr reviewed (changes requested):
Needs some fixes

mangose10 committed:
Update auth.controller.ts

mangose10 committed:
first set of review changes

first set only includes standard styling changes

mangose10 committed:
first set of reviews pt.2

small changes that were missed in the first commit

mangose10 committed:
second portion of RegisterForm review

This includes the password and email confirms.

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -9,6 +9,7 @@ import { DatabaseService } from '../database/database.service';
 import { User, UserCreateInput } from '../@generated/prisma-graphql/user';
 import { AuthResponse } from './auth-response';
 import { EmailService } from '../email/email.service';
+import { UniqueError } from "./uniqueError";
```
Don't need this

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -59,8 +57,8 @@ export class AuthService {
-   *          Success -> response.body will have user input data
-   *          Success, but user exists -> response.body will inncluse error {0,1,2}
+   *          Success -> valid JWT token
+   *          Fail, beacuse user exists -> response.body will include error {E_ERR, D_ERR, B_ERR} from UniqueError
```
Just use an HTTP error

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -59,8 +57,8 @@ export class AuthService {
+   *          Fail, beacuse user exists -> response.body will include error {E_ERR, D_ERR, B_ERR} from UniqueError
    */
   public async register(email: string, displayName: string, website: string = '', institution: string = '', password: string): Promise<{}> {
     var uci: UserCreateInput;
```
Don't use var in ES6. Use let. Don't abbreviate unless reaching code line length limit.

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -59,8 +57,8 @@ export class AuthService {
+   *          Success -> valid JWT token
+   *          Fail, beacuse user exists -> response.body will include error {E_ERR, D_ERR, B_ERR} from UniqueError
    */
   public async register(email: string, displayName: string, website: string = '', institution: string = '', password: string): Promise<{}> {
```
Don't need default values

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -69,10 +67,17 @@ export class AuthService {
+    if (true /*institution != ''*/){uci.institution = institution;}
 
-    return this.registerGraphQL(uci, password);
+    const user = this.registerGraphQL(uci, password);
```
Doesn't need a second separate function

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -114,26 +119,29 @@ export class AuthService {
-    let checkE = await this.$databaseService.user.findUnique({
-      where: {
-        email: input.email
+    try {
```
Doesn't need a try block

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -114,26 +119,29 @@ export class AuthService {
+        }
+      });
+      if (check?.length > 1){
+        throw UniqueError.B_ERR;
```
Use regular errors.

quinnjr commented on server/auth/uniqueError.ts:
```diff
@@ -0,0 +1,5 @@
+export enum UniqueError {
```
Unnecessary. Just use HTTP errors

quinnjr replied:
Typescript files are snake-case named

quinnjr commented on server/auth/not-unique.exception.ts:
```diff
@@ -1,15 +1,14 @@
  */
 export class NotUniqueException extends HttpException {
-  constructor(nui: any) {
+  constructor(nui: UniqueError) {
```
Unnecessary. Just use HTTP errors

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -114,26 +119,29 @@ export class AuthService {
-      where: {
-        displayName: input.displayName
+      if (check[0]?.email === input.email){
+        throw UniqueError.E_ERR;
```
Use HTTP errors

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -114,26 +119,29 @@ export class AuthService {
+        throw UniqueError.E_ERR;
+      }
+      if (check[0]?.displayName === input.displayName){
+        throw UniqueError.D_ERR;
```
Use HTTP errors

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -114,26 +119,29 @@ export class AuthService {
+        throw UniqueError.D_ERR;
+      }
+    }catch(e){
+      console.log(e);
```
remove console.log in production code

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -21,24 +20,23 @@ <h2>Register</h2>
               <p class="help is-danger">Must be a valid Email.</p>
             </div>
-            <div *ngIf="registerForm.controls.email.dirty && (!uniqueE)">
+            <div *ngIf="registerForm.controls.email.dirty && (!uniqueEmailFlag)">
```
Use angular forms methods to get dirty status. Use uniqueness error handling instead of a flag

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -21,24 +20,23 @@ <h2>Register</h2>
                 }" required />
             </div>
-            <div *ngIf="registerForm.controls.dName.dirty && registerForm.get('dName')?.hasError('required')">
+            <div *ngIf="registerForm.controls.displayName.dirty && registerForm.get('displayName')?.hasError('required')">
```
Use angular forms methods to get dirty status.

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -21,24 +20,23 @@ <h2>Register</h2>
               <p class="help is-danger">Display Name is required.</p>
             </div>
-            <div *ngIf="registerForm.get('dName')?.errors?.['InvalidDisplayName']">
+            <div *ngIf="registerForm.get('displayName')?.errors?.['InvalidDisplayName']">
```
use hasError

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -67,22 +65,23 @@ <h2>Register</h2>
-            <div *ngIf="registerForm.controls.password.dirty && (registerForm.get('password')?.errors?.['pattern'])">
-              <div>
-                <p class="help is-danger">Password must include atleast one of each:</p>
+            <div *ngIf="!registerForm.get('password')?.errors?.['required'] && !registerForm.get('password')?.valid" class="box">
```
use hasError

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -67,22 +65,23 @@ <h2>Register</h2>
-              <div>
-                <p class="help is-danger">Password must include atleast one of each:</p>
+            <div *ngIf="!registerForm.get('password')?.errors?.['required'] && !registerForm.get('password')?.valid" class="box">
+              <div *ngIf="(registerForm.get('password')?.errors?.['minlength'])">
```
use hasError

quinnjr commented on src/app/auth/register/register.component.html:
```diff
@@ -67,22 +65,23 @@ <h2>Register</h2>
+              <div *ngIf="(registerForm.get('password')?.errors?.['minlength'])">
+                <p class="help is-danger">Length of password must be at least 8.</p>
+              </div>
+              <div *ngIf="(registerForm.get('password')?.errors?.['pattern'])">
```
use hasError

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,33 +1,41 @@
 import { sameAsEmailValidator } from "./verify/validate/same-as-email.directive";
 import { samePassValidator } from "./verify/validate/pass-match.directive";
 import { HttpResponse, HttpClient } from '@angular/common/http';
+import { UniqueError } from "../../../../server/auth/uniqueError";
```
Unnecessary.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,33 +1,41 @@
-  public oldDName: string = '';
-  public uniqueE: boolean = true;
-  public uniqueD: boolean = true;
+  public oldDisplayName: string = '';
```
Unnecessary.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,33 +1,41 @@
-  public uniqueE: boolean = true;
-  public uniqueD: boolean = true;
+  public oldDisplayName: string = '';
+  public uniqueEmailFlag: boolean = true;
```
Unnecessary.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,33 +1,41 @@
-  public uniqueD: boolean = true;
+  public oldDisplayName: string = '';
+  public uniqueEmailFlag: boolean = true;
+  public uniqueDisplayNameFlag: boolean = true;
```
Unnecessary.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -40,20 +48,20 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
 
       this.oldEmail = this.registerForm.get('email')?.value.toLowerCase();
-      this.registerForm.controls['dName'].setValidators(
+      this.registerForm.controls['displayName'].setValidators(
```
Use get('displayName')

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -40,20 +48,20 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
-        this.registerForm.controls['dName'].reset();
-        this.registerForm.controls['dName'].setValue(this.registerForm.get('email')?.value.split("@", 2)[0]);
+      if (this.registerForm.get('email')?.value.split("@", 2)[0].toLowerCase() === this.registerForm.get('displayName')?.value.toLowerCase()){
+        this.registerForm.controls['displayName'].reset();
```
Don't array access. Use get('displayName')

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -40,20 +48,20 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
-        this.registerForm.controls['dName'].setValue(this.registerForm.get('email')?.value.split("@", 2)[0]);
+      if (this.registerForm.get('email')?.value.split("@", 2)[0].toLowerCase() === this.registerForm.get('displayName')?.value.toLowerCase()){
+        this.registerForm.controls['displayName'].reset();
+        this.registerForm.controls['displayName'].setValue(this.registerForm.get('email')?.value.split("@", 2)[0]);
```
Don't array access. Use get('displayName')

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -69,36 +77,31 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
-      institution:institution,
-      website:website,
-      password:password
+      email:email?.value.toLowerCase(),
```
Spaces with key: value pairs

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -69,36 +77,31 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
-      website:website,
-      password:password
+      email:email?.value.toLowerCase(),
+      displayName:displayName?.value,
```
Spaces with key: value pairs

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -69,36 +77,31 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
-      password:password
+      email:email?.value.toLowerCase(),
+      displayName:displayName?.value,
+      institution:institution?.value,
```
Spaces with key: value pairs

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -69,36 +77,31 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
+      email:email?.value.toLowerCase(),
+      displayName:displayName?.value,
+      institution:institution?.value,
+      website:website?.value,
```
Spaces with key: value pairs

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -69,36 +77,31 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
+      displayName:displayName?.value,
+      institution:institution?.value,
+      website:website?.value,
+      password:password?.value
```
Spaces with key: value pairs

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -69,36 +77,31 @@ export class RegisterComponent implements OnInit, OnDestroy, DoCheck {
-    });
+    },
+    err => {
+      if (err.error?.['error'].message === UniqueError.E_ERR){this.uniqueEmailFlag = false}
```
Use HTTP statuses

quinnjr reviewed (changes requested):
Needs a bit more work.

quinnjr commented on src/app/auth/register/verify/validate/same-as-email.directive.ts:
```diff
@@ -0,0 +1,14 @@
+ * @returns Valid -> null
+ *          Invalid -> InvalidDisplayName Error
+ */
+export function notEmailValidator(email: string): ValidatorFn {
```
public static matchValues(
    matchTo: string // name of the control to match to
  ): (AbstractControl) => ValidationErrors | null {
    return (control: AbstractControl): ValidationErrors | null => {
      return !!control.parent &&
        !!control.parent.value &&
        control.value === control.parent.controls[matchTo].value
        ? null
        : { isMatching: false };
    };
}

quinnjr replied:
https://stackoverflow.com/a/55929868

quinnjr replied:
This all also doesn't need to be in a separate file since we're not reusing the logic

mangose10 committed:
review pt. 3

still having troubles with squash, I can't manage to do it on the terminal or using git

quinnjr commented on server/auth/auth.controller.ts:
```diff
@@ -29,12 +30,9 @@ export class AuthController {
     @Body('password') password: string
   ): Promise<any> {
-    return this.$authService.register(email, displayName, website, institution, password);
+    return this.$authService.registerGraphQL(userInput, password);
```
This is not related to GraphQL

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -117,31 +82,28 @@ export class AuthService {
   public async registerGraphQL(
```
This is not related to GraphQL

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -117,31 +82,28 @@ export class AuthService {
+    });
+    //Email and Display name are in use
+    if (check?.length > 1){
+      throw new ForbiddenException(3);
```
This should be a string message, not a number value

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -117,31 +82,28 @@ export class AuthService {
+    }
+    //Email is in use
+    if (check[0]?.email === input.email){
+      throw new ForbiddenException(1);
```
This should be a string message, not a number value

quinnjr commented on server/auth/auth.service.ts:
```diff
@@ -117,31 +82,28 @@ export class AuthService {
+    }
+    //Display name is in use
+    if (check[0]?.displayName === input.displayName){
+      throw new ForbiddenException(2);
```
This should be a string message, not a number value

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -1,21 +1,17 @@
 import { HttpResponse, HttpClient } from '@angular/common/http';
-import { UniqueError } from "../../../../server/auth/uniqueError";
 import { AbstractControl, ValidatorFn, ValidationErrors } from "@angular/forms";
+import { UserCreateInput } from 'server/@generated/prisma-graphql/user';
```
This is not the correct import. Import it off 'prisma'.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -59,29 +56,13 @@ export class RegisterComponent implements OnInit, DoCheck {
+  /**
+   * Validates that display name is not email handle, case insensitive
+   */
+   notEmailHandleValidator: ValidatorFn = (group: AbstractControl):  ValidationErrors | null => {
```
Display name can be the user's email.

mangose10 replied:
I had understood that it did need to be different
![registerMemorandum](https://user-images.githubusercontent.com/23206954/172979209-c7b312e4-e0a6-4537-ab77-8dd056eb050c.PNG)

quinnjr replied:
Whoops. You're right.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -92,23 +73,30 @@ export class RegisterComponent implements OnInit, DoCheck {
+      'displayName' : displayName?.value,
+      'passwordHash':''
+    }
+    if (true /*website != ''*/){userInput.website = website?.value;}
```
Put on separate lines.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -92,23 +73,30 @@ export class RegisterComponent implements OnInit, DoCheck {
+      'passwordHash':''
+    }
+    if (true /*website != ''*/){userInput.website = website?.value;}
+    if (true /*institution != ''*/){userInput.institution = institution?.value;}
```
Put on separate lines.

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -92,23 +73,30 @@ export class RegisterComponent implements OnInit, DoCheck {
-      if (err.error?.['error'].message === UniqueError.E_ERR){this.uniqueEmailFlag = false}
-      if (err.error?.['error'].message === UniqueError.D_ERR){this.uniqueDisplayNameFlag = false}
-      if (err.error?.['error'].message === UniqueError.B_ERR){this.uniqueEmailFlag = false;this.uniqueDisplayNameFlag = false}
+      console.log(err)
```
Remove this

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -37,8 +33,9 @@ export class RegisterComponent implements OnInit, DoCheck {
-      passwordConfirm: ['', [Validators.required]]
-    }, {validators: [this.samePassValidator, this.sameEmailValidator]});
+      passwordConfirm: ['', [Validators.required]],
+      formStatus: [null]
```
This is not the correct way to display an error message

quinnjr commented on src/app/auth/register/register.component.ts:
```diff
@@ -92,23 +73,30 @@ export class RegisterComponent implements OnInit, DoCheck {
-      if (err.error?.['error'].message === UniqueError.B_ERR){this.uniqueEmailFlag = false;this.uniqueDisplayNameFlag = false}
+      console.log(err)
+      if (err.status === 403){
+        this.registerForm.get('formStatus')?.setValue(err.error.message)
```
This is not the correct way to display an error message.

Set another property on the class to handle error messages.

quinnjr committed:
Rebasing on current develop branch

quinnjr committed:
Resolving comment issues

quinnjr committed:
bugfix(): fixing compile errors

quinnjr committed:
Resolving comment issues, completed

quinnjr committed:
Merge branch 'develop' of github.com:quinnjr/pluma-online into feature/RegistrationForm
